api/pkg/s3/bucketsseput.go

```diff
@@ -0,0 +1,174 @@
+// Copyright 2019 The OpenSDS Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package s3
+
+import (
+	"crypto/md5"
+	"encoding/xml"
+	"fmt"
+	"github.com/opensds/multi-cloud/api/pkg/utils"
+	"net/http"
+	"strings"
+	"sync"
+
+	"github.com/emicklei/go-restful"
+	"github.com/opensds/multi-cloud/api/pkg/common"
+	. "github.com/opensds/multi-cloud/api/pkg/utils/constants"
+	"github.com/opensds/multi-cloud/s3/pkg/model"
+	s3 "github.com/opensds/multi-cloud/s3/proto"
+	log "github.com/sirupsen/logrus"
+)
+
+// Map from storage calss to tier
+var ClassAndTier1 map[string]int32
+var mutext1 sync.Mutex
+
+func (s *APIService) BucketSSEPut(request *restful.Request, response *restful.Response) {
+	bucketName := request.PathParameter("bucketName")
+	log.Infof("received request for PUT bucket SSE: %s", bucketName)
+
+	ctx := common.InitCtxWithAuthInfo(request)
+	bucket, err := s.s3Client.GetBucket(ctx, &s3.Bucket{Name: bucketName})
+	if err != nil {
+		log.Errorf("get bucket failed, err=%v\n", err)
+		response.WriteError(http.StatusInternalServerError, fmt.Errorf("bucket does not exist"))
+	}
+
+	body := ReadBody(request)
+	log.Infof("MD5 sum for body is %x", md5.Sum(body))
+
+	if body != nil {
+		sseConf := model.SSEConfiguration{}
+		errSSE := xml.Unmarshal(body, &sseConf)
+		if errSSE != nil {
+			response.WriteError(http.StatusInternalServerError, err)
+			return
+		}
+
+		s3SSE  := &s3.ServerSideEncryption{
+			SseType:              "",
+			EncryptionKey:        nil,
+			InitilizationVector:  nil,
+			XXX_NoUnkeyedLiteral: struct{}{},
+			XXX_unrecognized:     nil,
+			XXX_sizecache:        0,
+		}
+		if sseConf.SSE.Enabled=="true"{
+			s3SSE.SseType = "SSE"
+			s3SSE.EncryptionKey,_ = utils.GetRandom32BitKey()
+		}
+
+		bucket.BucketMeta.ServerSideEncryption = s3SSE
+
+		s.s3Client.UpdateBucket(ctx, bucket.BucketMeta)
```

Comment:
Hi @rhsakarpos , please think about the scenario that user create a bucket without SSE first, and set SSE later, it seems for this scenario, SSE can not be set successfully, cause nothing will change for UpdateBucket as there is no record in bucket_sseopts.

Reply:
Raised an issue #759 to track